# Muhammad Ali Zulfikar
## TI-3H
### 2141720088 / 20

Machine Learning Week #02

# PRAKTIKUM 1

In [131]:
import pandas as pd
from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn.model_selection import train_test_split

In [124]:
df = pd.read_csv('dataset/Titanic-Dataset.csv')
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [125]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


Well, bisa kita liat terdapat data yang berbeda jauh. Rata-rata data berjumlah 891, sedangkan ada beberapa data yang di bawah 891. Seperti **Embarked** atau **Cabin**, untuk memastikan asumsi apakah terdapat nilai null kita bisa melakukan pengecekan dengan 

```python
df.isnull().sum()
```

In [126]:
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

Sesuai data di atas, bisa dilihat bahwa 'Age' memiliki nilai NULL sebesar 177, 'Cabin' 687, dan 'Embarked' sebesar 2

In [129]:
df['Age'].fillna(value=df['Age'].mean(), inplace=True)
df['Cabin'].fillna(value="DECK", inplace=True)
df['Embarked'].fillna(value=df['Embarked'].mode, inplace=True)

Langkah di atas kita melakukan pengisian nilai yang memiliki NaN dengan rata-rata, modus, atau nilai yang sudah pasti

In [130]:
df.isnull().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64

# PRAKTIKUM 2

Selanjutnya, kita ingin melakukan analisis terhadap informasi penumpang yang tenggelam

In [113]:
df_sliced = df[['Survived', 'Pclass', 'Sex', 'Age', 'Cabin']]
df_sliced.head()

Survived  Pclass     Sex   Age Cabin
0         0       3    male  22.0  DECK
1         1       1  female  38.0   C85
2         1       3  female  26.0  DECK
3         1       1  female  35.0  C123
4         0       3    male  35.0  DECK

Kita perlu memberikan Encoding Label terhadap **Sex** dan **Cabin**

In [ ]:
encoder = LabelEncoder()
df_sliced.loc[:, 'Sex'] = encoder.fit_transform(df_sliced['Sex'])
df_sliced.loc[:, 'Cabin'] = encoder.fit_transform(df_sliced['Cabin'])

In [115]:
df_sliced.head()

Survived  Pclass  Sex   Age  Cabin
0         0       3    1  22.0    115
1         1       1    0  38.0     81
2         1       3    0  26.0    115
3         1       1    0  35.0     55
4         0       3    1  35.0    115

In [ ]:
std = StandardScaler()
df_sliced['Age'] = std.fit_transform(df_sliced['Age'].values.reshape(-1, 1))

In [117]:
df_sliced.head()

Survived  Pclass  Sex       Age  Cabin
0         0       3    1 -0.592481    115
1         1       1    0  0.638789     81
2         1       3    0 -0.284663    115
3         1       1    0  0.407926     55
4         0       3    1  0.407926    115

# Praktikum 3

### Random Split

In [132]:
df_train, df_unseen = train_test_split(df_sliced, test_size=0.2, random_state=0)
df_val, df_test = train_test_split(df_unseen, test_size=0.5, random_state=0)

In [133]:
print(f'Jumlah data asli: {df.shape[0]}')
print(f'Jumlah data train: {df_train.shape[0]}')
print(f'Jumlah data val: {df_val.shape[0]}')
print(f'Jumlah data test: {df_test.shape[0]}')

print('=========')
print(f'Jumlah label data asli:\n{df.Survived.value_counts()}')
print(f'Jumlah label data train:\n{df_train.Survived.value_counts()}')
print(f'Jumlah label data val:\n{df_val.Survived.value_counts()}')
print(f'Jumlah label data test:\n{df_test.Survived.value_counts()}')

Jumlah data asli: 891
Jumlah data train: 712
Jumlah data val: 89
Jumlah data test: 90
Jumlah label data asli:
0    549
1    342
Name: Survived, dtype: int64
Jumlah label data train:
0    439
1    273
Name: Survived, dtype: int64
Jumlah label data val:
0    53
1    36
Name: Survived, dtype: int64
Jumlah label data test:
0    57
1    33
Name: Survived, dtype: int64


### Stratified Split

In [136]:
df_st = df_sliced.copy()
df_st.head()

Survived  Pclass  Sex       Age  Cabin
0         0       3    1 -0.592481    115
1         1       1    0  0.638789     81
2         1       3    0 -0.284663    115
3         1       1    0  0.407926     55
4         0       3    1  0.407926    115

In [137]:
df_st_train, df_st_unseen = train_test_split(df_st, test_size=0.2, random_state=0, stratify=df_st['Survived'])
df_st_val, df_st_test = train_test_split(df_st_unseen, test_size=0.2, random_state=0, stratify=df_st_unseen['Survived'])

In [139]:
print(f'Jumlah label data asli:\n{df_st_train.Survived.value_counts()}')
print(f'Jumlah label data train:\n{df_st_unseen.Survived.value_counts()}')
print(f'Jumlah label data val:\n{df_st_val.Survived.value_counts()}')
print(f'Jumlah label data test:\n{df_st_test.Survived.value_counts()}')

Jumlah label data asli:
0    439
1    273
Name: Survived, dtype: int64
Jumlah label data train:
0    110
1     69
Name: Survived, dtype: int64
Jumlah label data val:
0    88
1    55
Name: Survived, dtype: int64
Jumlah label data test:
0    22
1    14
Name: Survived, dtype: int64


In [140]:
df_split = df_sliced.copy()
df_split.head()

Survived  Pclass  Sex       Age  Cabin
0         0       3    1 -0.592481    115
1         1       1    0  0.638789     81
2         1       3    0 -0.284663    115
3         1       1    0  0.407926     55
4         0       3    1  0.407926    115

In [141]:
from sklearn.model_selection import KFold
kf = KFold(n_splits=4)
print(f'Jumlah fold: {kf.get_n_splits()}')
print(f'Obyek KFold: {kf}')

Jumlah fold: 4
Obyek KFold: KFold(n_splits=4, random_state=None, shuffle=False)


In [143]:
kf_split = kf.split(df_split)
print(f'Jumlah data df: {df_split.shape[0]}')
for train_index, test_index in kf_split:
    print(f'Index train: {train_index} | Index test: {test_index}')

Jumlah data df: 891
Index train: [223 224 225 226 227 228 229 230 231 232 233 234 235 236 237 238 239 240
 241 242 243 244 245 246 247 248 249 250 251 252 253 254 255 256 257 258
 259 260 261 262 263 264 265 266 267 268 269 270 271 272 273 274 275 276
 277 278 279 280 281 282 283 284 285 286 287 288 289 290 291 292 293 294
 295 296 297 298 299 300 301 302 303 304 305 306 307 308 309 310 311 312
 313 314 315 316 317 318 319 320 321 322 323 324 325 326 327 328 329 330
 331 332 333 334 335 336 337 338 339 340 341 342 343 344 345 346 347 348
 349 350 351 352 353 354 355 356 357 358 359 360 361 362 363 364 365 366
 367 368 369 370 371 372 373 374 375 376 377 378 379 380 381 382 383 384
 385 386 387 388 389 390 391 392 393 394 395 396 397 398 399 400 401 402
 403 404 405 406 407 408 409 410 411 412 413 414 415 416 417 418 419 420
 421 422 423 424 425 426 427 428 429 430 431 432 433 434 435 436 437 438
 439 440 441 442 443 444 445 446 447 448 449 450 451 452 453 454 455 456
 457 458 459 460 4

# Praktikum 4

In [144]:
from PIL import Image

img = Image.open('dataset/Lenna_(test_image).png')
img.show()

In [145]:
r, g, b = img.split()
print(len(r.histogram()))
print(r.histogram())

256
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 8, 6, 10, 9, 20, 37, 32, 41, 53, 74, 103, 132, 179, 193, 242, 238, 336, 386, 404, 483, 607, 628, 771, 785, 1001, 1016, 1177, 1269, 1332, 1446, 1487, 1457, 1574, 1585, 1591, 1557, 1569, 1660, 1648, 1420, 1559, 1418, 1384, 1319, 1342, 1156, 1120, 955, 969, 828, 782, 752, 737, 719, 700, 628, 673, 587, 617, 610, 592, 557, 593, 552, 566, 582, 559, 571, 520, 664, 650, 618, 730, 594, 667, 675, 685, 771, 715, 667, 740, 744, 766, 765, 772, 817, 817, 744, 806, 760, 777, 812, 797, 799, 861, 814, 910, 907, 918, 888, 1011, 879, 996, 912, 952, 884, 1074, 977, 1073, 1040, 1216, 1250, 1403, 1534, 1639, 1682, 1776, 1874, 1769, 1582, 1743, 1441, 1477, 1483, 1409, 1437, 1449, 1389, 1479, 1592, 1655, 1657, 1666, 1857, 1896, 1813, 1979, 1814, 1956, 1928, 2055, 2012, 2303, 2333, 2670, 2787, 3232, 3154, 3476, 3424, 3516, 3102, 31